### PR TITLE
Verifiable smart contract builds.

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Add notes about verifiable builds.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by dg-concordium — APPROVED** [review]

File: source/mainnet/smart-contracts/guides/compile-module.rst [hunk 1]
```diff
@@ -30,14 +30,31 @@ To help building smart contract modules and to take advantage of features
...
+   $cargo concordium build --verifiable docker.io/concordium/verifiable-sc:1.70.0 --out contract.wasm.v1
 
 This uses Cargo_ for building, but runs further optimizations on the result.
 
+The ``--verifiable`` option will cause ``cargo-concordium`` to build the sources
+in a fixed container to make the build reproducible. It will additional produce
+a ``tar`` file with packaged sources that were used for the build. The list of
+available docker images can be found on `DockerHub. <https://hub.docker.com/r/concordium/verifiable-sc>`_
+
+This option is available on ``cargo-concordium`` version 3.1 or newer.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply source/mainnet/smart-contracts/guides/compile-module.rst]: Removed

File: source/mainnet/smart-contracts/guides/compile-module.rst [hunk 2]
```diff
@@ -30,14 +30,29 @@ To help building smart contract modules and to take advantage of features
...
+a ``tar`` file with packaged sources that were used for the build. The list of
+available docker images can be found on `DockerHub. <https://hub.docker.com/r/concordium/verifiable-sc>`_
+
+The ``tar`` archive should be uploaded and made publicly available, and its
+link should be embedded into the deployed module using the ``cargo concordium edit-build-info``
+command.
+
+The ``--out`` option is required for verifiable builds, and instructs ``cargo-concordium`` to output the build artifact at the specified location.
+
+Verifiable builds require a container runtime such as `Docker <https://www.docker.com/>`_ to be installed.
```
Comment (limemloh) [inline source/mainnet/smart-contracts/guides/compile-module.rst:52]: This should probably also be mentioned in the `setup-tools` guide, that we link in the preparation section

**Review by mh-concordium — APPROVED** [review]
Looks good.

File: source/mainnet/smart-contracts/guides/compile-module.rst [hunk 3]
```diff
@@ -30,14 +30,29 @@ To help building smart contract modules and to take advantage of features
...
 This uses Cargo_ for building, but runs further optimizations on the result.
 
+The ``--verifiable`` option  is available on ``cargo-concordium`` version 3.1 or newer. It will cause ``cargo-concordium`` to build the sources
+in a fixed container to make the build reproducible. It will additionally produce
+a ``tar`` file with packaged sources that were used for the build. The list of
+available docker images can be found on `DockerHub. <https://hub.docker.com/r/concordium/verifiable-sc>`_
+
+The ``tar`` archive should be uploaded and made publicly available, and its
+link should be embedded into the deployed module using the ``cargo concordium edit-build-info``
+command.
```
Comment (limemloh) [inline source/mainnet/smart-contracts/guides/compile-module.rst:48]: Consider adding an example of the command being called with some dummy value

File: source/mainnet/smart-contracts/guides/compile-module.rst [hunk 4]
```diff
@@ -73,11 +89,17 @@ If you want a smaller code footprint, enable the :ref:`wee_alloc feature <setup-
    Note that even with ``--release`` set, the produced Wasm module includes
    debug information.
 
-Running the ``cargo concordium build`` command will produce a smart contract module which can be found
+Non-verifiable builds
+---------------------
+
+Omitting the ``--verifiable`` option from the build command will make a build on the host machine.
+This is typically not verifiable since the Rust toolchain is affected by the build environment.
+However, for local development, non-verifiable builds can be useful.
```
Comment (limemloh) [inline source/mainnet/smart-contracts/guides/compile-module.rst:97]: I guess closed sourced smart contracts will also want to avoid using it.

**Review by limemloh — APPROVED** [review]

File: source/mainnet/smart-contracts/guides/compile-module.rst (hunk 3 above)
Comment (abizjak) [reply source/mainnet/smart-contracts/guides/compile-module.rst:48]: Added

File: source/mainnet/smart-contracts/guides/compile-module.rst (hunk 4 above)
Comment (abizjak) [reply source/mainnet/smart-contracts/guides/compile-module.rst:97]: Yes, but I don't want to mention those. It's fundamentally at odds with decentralization to do things on the blockchain and keep the source private .

File: source/mainnet/smart-contracts/guides/compile-module.rst (hunk 2 above)
Comment (abizjak) [reply source/mainnet/smart-contracts/guides/compile-module.rst:52]: Added

**Pull request merged**